src/mgr/ActivePyModules.cc

```diff
@@ -876,7 +880,77 @@ PyObject* ActivePyModules::with_perf_counters(
   return f.get();
 }
 
-PyObject* ActivePyModules::get_counter_python(
+PyObject* ActivePyModules::with_perf_counters(
+    std::function<void(PerfCounterInstance& counter_instance, PerfCounterType& counter_type, PyFormatter& f)> fct,
+    const std::string &svc_name,
+    const std::string &svc_id,
+    const std::string_view &counter_name,
+    const std::string_view &sub_counter_name,
+    const std::vector<std::pair<std::string_view,std::string_view>> &labels) const
+{
+  PyFormatter f;
+  /*
+    The resolved counter path, they are of the format
+    <counter_name>.<sub_counter_name> If the counter name has labels, then they
+    are segregated via NULL delimters.
+
+    Eg:
+      - labeled counter:
+        "osd_scrub_sh_repl^@level^@shallow^@pooltype^@replicated^@.successful_scrubs_elapsed"
+      - unlabeled counter: "osd.stat_bytes"
+  */
+  std::string resolved_path;
+  // NOTE: This should ideally be the resolved_path, but since the resolved path
+  // contains NULL delimiters and we cannot have an array section whose key has
+  // NULL characters, we will just use the counter_name to extract the value at
+  // the call site
+  f.open_array_section(counter_name);
+
+  // Construct the resolved path
+  if (!labels.empty()){
+    // NOTE: The 'ceph::perf_counters::key_create()' API requires a static
+    // array. Hence we are assuming that a performance counter will never have
+    // more than 100 labels.
+    const int max_labels_per_key = 100;
+    ceph::perf_counters::label_pair labels_array[max_labels_per_key];
+    std::copy(labels.begin(), labels.end(), std::begin(labels_array));
+    std::string counter_name_with_labels = ceph::perf_counters::key_create(counter_name.data(), std::move(labels_array));
+    resolved_path = std::string(counter_name_with_labels) + "." + std::string(sub_counter_name);
+  } else {
+    resolved_path = std::string(counter_name) + "." + std::string(sub_counter_name);
+  }
+
+
+  {
+    without_gil_t no_gil;
+    std::lock_guard l(lock);
+    auto metadata = daemon_state.get(DaemonKey{svc_name, svc_id});
+    if (metadata) {
+      std::lock_guard l2(metadata->lock);
+      if (metadata->perf_counters.instances.count(resolved_path)) {
+        auto counter_instance = metadata->perf_counters.instances.at(resolved_path);
+        auto counter_type = metadata->perf_counters.types.at(resolved_path);
+        with_gil(no_gil, [&] {
+          fct(counter_instance, counter_type, f);
+        });
+      } else {
+        dout(4) << "Missing counter: '" << resolved_path << "' ("
+		<< svc_name << "." << svc_id << ")" << dendl;
+        dout(20) << "Paths are:" << dendl;
+        for (const auto &i : metadata->perf_counters.instances) {
+          dout(20) << i.first << dendl;
+        }
+      }
+    } else {
+      dout(4) << "No daemon state for " << svc_name << "." << svc_id << ")"
+              << dendl;
+    }
+  }
+  f.close_section();
+  return f.get();
```

Comment:
That last line is correct? please verify

Reply:
yes, this is correct. `with_perf_counters` returns a `PyObject` where the `key` is the counter name and the `value` is the value of the counter. And `f.get()` closes the formatter stream and construct the `PyObject` that we want to return